src/FsAutoComplete.Core/AdaptiveExtensions.fs

```diff
@@ -288,3 +291,517 @@ module AMap =
+  /// of other AdaptiveObjects since it does not track dependencies.
+  /// </summary>
+  /// <remarks>
+  /// This follows Task semantics and is already running.
+  /// </remarks>
+  let force (value: asyncaval<_>) = value.GetValue(AdaptiveToken.Top)
+
+  /// <summary>
+  /// Evaluates the given adaptive value and returns an Async containing the value.
+  /// This should not be used inside the adaptive evaluation
+  /// of other AdaptiveObjects since it does not track dependencies.
+  /// </summary>
+  /// <remarks>
+  /// This follows Async semantics and is not already running.
+  /// </remarks>
+  let forceAsync (value: asyncaval<_>) =
+    async {
+      let ct = value.GetValue(AdaptiveToken.Top)
+      return! Async.ofAdaptiveCancellableTask ct
+    }
+
+  /// <summary>
+  /// Evaluates the given adaptive value and returns a CancellableTask containing the value.
+  /// This should not be used inside the adaptive evaluation
+  /// of other AdaptiveObjects since it does not track dependencies.
+  /// </summary>
+  /// <remarks>
+  /// This follows CancellableTask semantics and is not already running.
+  /// </remarks>
+  let forceCancellableTask (value: asyncaval<_>) =
+    cancellableTask {
+      let ct = value.GetValue(AdaptiveToken.Top)
+      return! ct
+    }
+
+  /// A constant value that results in a Task.
+  type ConstantVal<'a>(value: AdaptiveCancellableTask<'a>) =
+    inherit ConstantObject()
+
+    new(value: Task<'a>) = ConstantVal<'a>(AdaptiveCancellableTask(id, value))
+
+    interface asyncaval<'a> with
+      member x.GetValue _ = value
+
+  /// <summary>
+  /// Base class for standard Async Adaptive Values.
+  /// </summary>
+  [<AbstractClass>]
+  type AbstractVal<'a>() =
+    inherit AdaptiveObject()
+    abstract Compute: AdaptiveToken -> AdaptiveCancellableTask<'a>
+
+    member x.GetValue token = x.EvaluateAlways token x.Compute
+
+    interface asyncaval<'a> with
+      member x.GetValue t = x.GetValue t
+
+  /// <summary>
+  /// Creates a constant async adaptive value always holding the given value.
+  /// </summary>
+  let constant (value: 'a) =
+    ConstantVal(Task.FromResult value) :> asyncaval<_>
+
+  /// <summary>
+  /// Creates a constant async adaptive value always holding the task.
+  /// </summary>
+  let ofTask (value: Task<'a>) = ConstantVal(value) :> asyncaval<_>
+
+  // let ofCancelableTask (value: AdaptiveCancellableTask<'a>) = ConstantVal(value) :> asyncaval<_>
+
+  // let ofCancellableTask (value : CancellableTask<'a>) =
+  //     ConstantVal (
+  //         let cts = new CancellationTokenSource ()
+  //         let cancel () =
+  //             cts.Cancel()
+  //             cts.Dispose()
+  //         let real = task {
+  //             try
+  //                 return! value cts.Token
+  //             finally
+  //                 cts.Dispose()
+  //         }
+  //         AdaptiveCancellableTask(cancel, real)
+  //     )
+  //     :> asyncaval<_>
```

Comment:
There are a couple of conversions I have yet to create because I get the feeling they may not work exactly as intended. May just remove these.